File: leetcode/Matrix/0289_Game_of_Life.py
class Solution(object):
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        N, M = len(board), len(board[0])
        neighbors = [[0] * M for _ in range(N)]
        for i in range(N):
            for j in range(M):
                if i > 0 and j > 0:
                    neighbors[i][j] += board[i-1][j-1]
                if i > 0:
                    neighbors[i][j] += board[i-1][j]
                if i > 0 and j < M - 1:
                    neighbors[i][j] += board[i-1][j+1]
                if j < M - 1:
                    neighbors[i][j] += board[i][j+1]
                if i < N - 1 and j < M - 1:
                    neighbors[i][j] += board[i+1][j+1]
                if i < N - 1:
                    neighbors[i][j] += board[i+1][j]
                if i < N - 1 and j > 0:
                    neighbors[i][j] += board[i+1][j-1]
                if j > 0:
                    neighbors[i][j] += board[i][j-1]

        for i in range(N):
            for j in range(M):
                if board[i][j] == 1:
                    if neighbors[i][j] < 2 or neighbors[i][j] > 3:
                        board[i][j] = 0
                else:
                    if neighbors[i][j] == 3:
                        board[i][j] = 1

File: leetcode/Matrix/0289_Game_of_Life.cpp
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        const int live = 1, dead = 0, liveToDead = 2, deadToLive = -1;
        int n = board.size(), m = board[0].size();
        
        for (int i = 0; i < n; i++) {
        	for (int j = 0; j < m; j++) {
        		// calculate live neighbors
        		int liveNeighbors = 0;
        		if (i != 0 && (board[i - 1][j] == live || board[i - 1][j] == liveToDead))
        			liveNeighbors ++;
        		if (i != 0 && j != 0 && (board[i - 1][j - 1] == live || board[i - 1][j - 1] == liveToDead))
        			liveNeighbors ++;
        		if (j != 0 && (board[i][j - 1] == live || board[i][j - 1] == liveToDead))
        			liveNeighbors ++;
        		if (i != n - 1 && j != 0 && (board[i + 1][j - 1] == live || board[i + 1][j - 1] == liveToDead))
        			liveNeighbors ++;
        		if (i != n - 1 && (board[i + 1][j] == live || board[i + 1][j] == liveToDead))
        			liveNeighbors ++;
        		if (i != n - 1 && j != m - 1 && (board[i + 1][j + 1] == live || board[i + 1][j + 1] == liveToDead))
        			liveNeighbors ++;
        		if (j != m - 1 && (board[i][j + 1] == live || board[i][j + 1] == liveToDead))
        			liveNeighbors ++;
        		if (i != 0 && j != m - 1 && (board[i - 1][j + 1] == live || board[i - 1][j + 1] == liveToDead))
        			liveNeighbors ++;

        		if (board[i][j] == 1 && (liveNeighbors < 2 || liveNeighbors > 3)) {
        			board[i][j] = liveToDead;
        		}
        		else if (board[i][j] == 0 && liveNeighbors == 3){
        			board[i][j] = deadToLive;
        		}
        	}
        }

        for (int i = 0; i < n; i++) {
        	for (int j = 0; j < m; j++) {
        		if (board[i][j] == liveToDead)
        			board[i][j] = dead;
        		else if (board[i][j] == deadToLive)
        			board[i][j] = live;
        	}
        }
    }
};